File: Daily/PY/Leetcode166-分数到小数.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        if numerator % denominator == 0: return str(numerator // denominator)
        up,down = abs(numerator),abs(denominator)
        ans = "-" if (numerator < 0) ^ (denominator < 0) else ''
        ans += str(up // down) + '.'
        st = dict()
        up = up % down * 10
        while up != 0:
            if up in st:
                ans = ans[:st[up]] + '(' + ans[st[up]:]
                ans += ')'
                break
            st[up] = len(ans)
            ans += str(up // down)
            up = up % down * 10
        return ans

File: Leetcode/sbgo/Leetcode166-分数到小数.go
package sbgo

import "strconv"

func fractionToDecimal(numerator int, denominator int) string {
	if numerator%denominator == 0 {
		return strconv.Itoa(numerator / denominator)
	}
	var ans string
	if (numerator < 0) != (denominator < 0) {
		ans = "-"
	}
	up, down := abs(numerator), abs(denominator)
	// 整数
	ans += strconv.Itoa(up/down) + "."
	// 小数部分
	st := map[int]int{}
	up = up % down * 10
	for up != 0 {
		if _, ok := st[up]; ok {
			ans = ans[:st[up]] + "(" + ans[st[up]:]
			ans += ")"
			break
		}
		st[up] = len(ans)
		ans += strconv.Itoa(up / down)
		up = up % down * 10
	}
	return ans
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
